lib/publiccloud/instance.pm

```diff
@@ -477,4 +481,202 @@ sub network_speed_test() {
     record_info("curl $rmt_host", $self->run_ssh_command(cmd => "curl -w '$write_out' -o /dev/null -v https://$rmt_host/", proceed_on_failure => 1));
 }
 
+=head2 measure_boottime
+
+    measure_boottime();
+
+Perfomrance measurement of the system Boot time. 
+Mainly used C<systemd-analyze> command for the data extraction.
+Data is then collected in an internal record, ready for storing in a DB.
+- when PUBLIC_CLOUD_PERF_COLLECT=0 boottime measurement disabled;
+- when bit-0 ON,First reboot only enabled =1;
+- when bit-1 ON, first,softreboot enabled =3 [2 too];
+- when bit-2 ON, first,hardreboot enabled =5 [4 too];
+- when all bits ON,all 3 measures enabled =7 [6 too].
+
+=cut
+
+sub measure_boottime() {
+    my ($self, $instance) = @_;
+    my $perf = get_var('PUBLIC_CLOUD_PERF_COLLECT');
+    return 0 unless ($perf);
+
+    my $softreboot = $perf & 2;    # bit-1 check
+    my $hardreboot = $perf & 4;    # bit-2 check
+
+    my $ret = {
+        kernel_release => undef,
+        kernel_version => undef,
+        analyze => {},
+        blame => {
+            first => {}, soft => {}, hard => {}
+        },
+    };
+    record_info("BOOT TIME", 'systemd_analyze');
+    # first deployment analysis
+    my ($systemd_analyze, $systemd_blame) = do_systemd_analyze($instance);
+    $ret->{analyze}->{$_} = $systemd_analyze->{$_} foreach (keys(%{$systemd_analyze}));
+    $ret->{blame}->{first} = $systemd_blame;
+
+    # Collect kernel version
+    $ret->{kernel_release} = $instance->run_ssh_command(cmd => 'uname -r');
+    $ret->{kernel_version} = $instance->run_ssh_command(cmd => 'uname -v');
+
+    if ($softreboot) {
+        # Do soft reboot
+        my ($shutdown_time, $startup_time) = $instance->softreboot();
+        $ret->{analyze}->{ssh_access_soft} = $startup_time;
+
+        ($systemd_analyze, $systemd_blame) = do_systemd_analyze($instance);
+        $ret->{analyze}->{$_ . '_soft'} = $systemd_analyze->{$_} foreach (keys(%{$systemd_analyze}));
+        $ret->{blame}->{soft} = $systemd_blame;
+    }
+
+    if ($hardreboot) {
+        # Do hard reboot
+        $instance->stop();
+        $ret->{analyze}->{ssh_access_hard} = $instance->start();
+
+        ($systemd_analyze, $systemd_blame) = do_systemd_analyze($instance);
+        $ret->{analyze}->{$_ . '_hard'} = $systemd_analyze->{$_} foreach (keys(%{$systemd_analyze}));
+        $ret->{blame}->{hard} = $systemd_blame;
+    }
+
+    # Do logging to openqa UI
+    $Data::Dumper::Sortkeys = 1;
+    record_info("RESULTS", Dumper($ret));
+
+    $instance->run_ssh_command(cmd => 'sudo tar -czvf /tmp/sle_cloud.tar.gz /var/log/cloudregister /var/log/cloud-init.log /var/log/cloud-init-output.log /var/log/messages /var/log/NetworkManager', proceed_on_failure => 1);
+    $instance->upload_log('/tmp/sle_cloud.tar.gz');
```

Comment:
upload log also has proceed_on_failure key , we should make this code transparent to all other setup . we don't want ltp test to fail because it can not measure boot time

Reply:
I set the parameter `failok => 1`, as in [upload_logs](https://open.qa/api/testapi/#_upload_logs)